### PR TITLE
feat: implement 'argocd admin appset generate' to troubeshoot appsets

### PR DESCRIPTION
**Pull request by alexmt**
PR implements `argocd appset generate` that helps troubleshooting appset issues. Examples:

```bash
argocd appset generate ~/go/src/github.com/argoproj/argo-cd/applicationset/examples/list-generator/list-
```

<details>
  <summary>output</summary>
  

``` yaml
- apiVersion: argoproj.io/v1alpha1
  kind: Application
  metadata:
    finalizers:
    - resources-finalizer.argocd.argoproj.io
    name: engineering-dev-guestbook
  spec:
    destination:
      namespace: guestbook
      server: https://kubernetes.default.svc
    project: default
    source:
      path: applicationset/examples/list-generator/guestbook/engineering-dev
      repoURL: https://github.com/argoproj/argo-cd.git
      targetRevision: HEAD
- apiVersion: argoproj.io/v1alpha1
  kind: Application
  metadata:
    finalizers:
    - resources-finalizer.argocd.argoproj.io
    name: engineering-prod-guestbook
  spec:
    destination:
      namespace: guestbook
      server: https://kubernetes.default.svc
    project: default
    source:
      path: applicationset/examples/list-generator/guestbook/engineering-prod
      repoURL: https://github.com/argoproj/argo-cd.git
      targetRevision: HEAD
```

</details>

```bash
argocd appset generate ~/go/src/github.com/argoproj/argo-cd/applicationset/examples/git-generator-directory/git-directories-example.yaml
```

<details>
  <summary>output</summary>
  

``` yaml
- apiVersion: argoproj.io/v1alpha1
  kind: Application
  metadata:
    finalizers:
    - resources-finalizer.argocd.argoproj.io
    name: argo-workflows
  spec:
    destination:
      namespace: argo-workflows
      server: https://kubernetes.default.svc
    project: my-project
    source:
      path: applicationset/examples/git-generator-directory/cluster-addons/argo-workflows
      repoURL: https://github.com/argoproj/argo-cd.git
      targetRevision: HEAD
    syncPolicy:
      syncOptions:
      - CreateNamespace=true
- apiVersion: argoproj.io/v1alpha1
  kind: Application
  metadata:
    finalizers:
    - resources-finalizer.argocd.argoproj.io
    name: prometheus-operator
  spec:
    destination:
      namespace: prometheus-operator
      server: https://kubernetes.default.svc
    project: my-project
    source:
      path: applicationset/examples/git-generator-directory/cluster-addons/prometheus-operator
      repoURL: https://github.com/argoproj/argo-cd.git
      targetRevision: HEAD
    syncPolicy:
      syncOptions:
      - CreateNamespace=true
```

</details>

```bash
argocd appset generate ~/go/src/github.com/argoproj/argo-cd/applicationset/examples/pull-request-generator/pull-request-example.yaml
```

<details>
  <summary>failed output</summary>
  

```
ERRO[0000] error generating params                       error="failed to select pull request service provider: error fetching Secret token: error fetching secret /github-token: an empty namespace may not be set when a resource name is provided" generator="&{0x1400128d170 0x104833710 { [] true 0x1400126b1d0}}"
ERRO[0000] error generating application from params      error="failed to select pull request service provider: error fetching Secret token: error fetching secret /github-token: an empty namespace may not be set when a resource name is provided" generator="{nil nil nil nil nil &PullRequestGenerator{Github:&PullRequestGeneratorGithub{Owner:myorg,Repo:myrepo,API:https://git.example.com/,TokenRef:&SecretRef{SecretName:github-token,Key:token,},AppSecretName:,Labels:[preview],},GitLab:nil,Gitea:nil,BitbucketServer:nil,Filters:[]PullRequestGeneratorFilter{},RequeueAfterSeconds:nil,Template:ApplicationSetTemplate{ApplicationSetTemplateMeta:ApplicationSetTemplateMeta{Name:,Namespace:,Labels:map[string]string{},Annotations:map[string]string{},Finalizers:[],},Spec:ApplicationSpec{Source:nil,Destination:ApplicationDestination{Server:,Namespace:,Name:,},Project:,SyncPolicy:nil,IgnoreDifferences:[]ResourceIgnoreDifferences{},Info:[]Info{},RevisionHistoryLimit:nil,Sources:[]ApplicationSource{},},},Bitbucket:nil,AzureDevOps:nil,} nil nil nil nil}"
Error: failed to select pull request service provider: error fetching Secret token: error fetching secret /github-token: an empty namespace may not be set when a resource name is provided
```

</details>

### CODE REVIEW

**Review by todaywasawesome — COMMENTED** [review]
@alexmt Part of the implementation here is with an embedded repo server, can you share how that works?

**Comment by daftping** [comment]
I wonder how this command differs from a recently merged similar PR.
https://github.com/argoproj/argo-cd/pull/16781

File: cmd/argocd/commands/headless/headless.go [hunk 1]
```diff
@@ -109,6 +117,86 @@ func (c *forwardCacheClient) NotifyUpdated(key string) error {
...
+
+func (c *embeddedRepoClientset) UpdateRevisionForPaths(ctx context.Context, in *repoapiclient.UpdateRevisionForPathsRequest, opts ...grpc.CallOption) (*repoapiclient.UpdateRevisionForPathsResponse, error) {
+	return c.service.UpdateRevisionForPaths(ctx, in)
+}
+
+func (c *embeddedRepoClientset) NewRepoServerClient() (io.Closer, repoapiclient.RepoServerServiceClient, error) {
+	return io.NopCloser, c, nil
+}
+
+func newEmbeddedRepoClientset() repoapiclient.Clientset {
```
Comment (pasha-codefresh) [inline cmd/argocd/commands/headless/headless.go:191]: i think we need to pass opts here as well, for future reusability.

File: cmd/argocd/commands/headless/headless.go [hunk 2]
```diff
@@ -109,6 +117,86 @@ func (c *forwardCacheClient) NotifyUpdated(key string) error {
 	})
 }
 
+func NewRepoClientset(namespace string, ctxStr string, clientOpts *apiclient.ClientOptions, kubeClientset *kubernetes.Clientset) repoapiclient.Clientset {
```
Comment (pasha-codefresh) [inline cmd/argocd/commands/headless/headless.go:120]: Maybe split it somehow, because for embedded, you actually dont need all these params, and it may look a bit strange when you creating it

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]
Amazing feature. Thank you Alex, just few small comments

File: cmd/argocd/commands/admin/appset.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+func NewAppsetCommand(opts *apiclient.ClientOptions) *cobra.Command {
```
Comment (agaudreault) [inline cmd/argocd/commands/admin/appset.go:33]: I feel like this does not make a lot of sense under the admin commands. Maybe `argocd appset template ./appset.yaml -o yaml` would make more sense. Take a look at https://github.com/argoproj/argo-cd/pull/16781, that has prior work to implement this feature.

File: cmd/argocd/commands/root.go [hunk 4]
```diff
@@ -85,6 +85,7 @@ func NewCommand() *cobra.Command {
...
 	command.PersistentFlags().StringVar(&clientOpts.RedisName, "redis-name", env.StringFromEnv(common.EnvRedisName, common.DefaultRedisName), fmt.Sprintf("Name of the Redis deployment; set this or the %s environment variable when the Redis's name label differs from the default, for example when installing via the Helm chart", common.EnvRedisName))
 	command.PersistentFlags().StringVar(&clientOpts.RepoServerName, "repo-server-name", env.StringFromEnv(common.EnvRepoServerName, common.DefaultRepoServerName), fmt.Sprintf("Name of the Argo CD Repo server; set this or the %s environment variable when the server's name label differs from the default, for example when installing via the Helm chart", common.EnvRepoServerName))
+	command.PersistentFlags().BoolVar(&clientOpts.EmbeddedRepoServer, "embedded-repo-server", env.ParseBoolFromEnv(common.EnvEmbeddedRepoServer, false), fmt.Sprintf("Indicates that 'argocd' CLI should run in-process repo server with default settings; set this or the %s environment variable", common.EnvEmbeddedRepoServer))
```
Comment (agaudreault) [inline cmd/argocd/commands/root.go:88]: I think this should be in it's own PR.

**Review by agaudreault — CHANGES_REQUESTED** [review]

**Comment by alexmt** [comment]
@agaudreault, @daftping, Thanks for letting me know about `argocd appset create --dry-run` https://github.com/argoproj/argo-cd/pull/16781! I did not know it was implemented.

My main goal is to enable customers to troubleshoot application sets. Pros of `argocd admin appset generate`:

* Command prints all related logs if app generation fails - currently we often have to check applicationset controller logs to understand what went wrong. ApplicationSet status is not enough
* Command prints YAML of generated applications - this enables user verify that appset is working as expected
* Command allows using local git repos using `file:///<path>/<to>/<repo>` (embedded repo server)

Cons:

* `argocd admin appset generate` works only for admins, while `argocd admin appset generate` works for everyone
* the `argocd admin appset generate` is already there so it is confusing to introduce another command


I propose to improve `argocd appset create --dry-run`  to return error logs (if any ) and the list of generate argocd apps. We won't be able to use local git repos but it is anyway nice to have feature. If no objections I would close this PR and work on `argocd appset create --dry-run` improvements in another PR.

**Comment by agaudreault** [comment]
@alexmt sounds good to me! Last time I looked at it, I chose not to return the full templated app  so i wouldn't change the return type of the create. By default, it won't return the template when it is not a dry run because they are generated async. I didn't want to have the create returns false results.

I think it would be better to introduce a new template command that would return the appset and the templates of generated apps. And if a name is provided instead of a file, it returns the actual data.

That was my state of mind as far as i can recall, but I'll let you judge what's best when you get in the code!

**Comment by kencochrane** [comment]
@alexmt I think closing this one and adding improvement to `argocd appset create --dry-run` make sense. Thank you

File: cmd/argocd/commands/admin/appset.go (hunk 3 above)
Comment (alexmt) [reply cmd/argocd/commands/admin/appset.go]: Agree - moved command to `argocd appset generate`. I've choosen to name it `generate` , however don't have strong opinion . Please let me know what you think

File: cmd/argocd/commands/admin/appset.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+		clientConfig clientcmd.ClientConfig
+		outputFormat string
+		debug        bool
+	)
+	command := &cobra.Command{
+		Use:   "generate",
+		Short: "Generate applications for given ApplicationSet",
+		Run: func(c *cobra.Command, args []string) {
+			if !debug {
+				log.SetLevel(log.WarnLevel)
```
[the remark this answers is not in the document]
Comment (alexmt) [reply cmd/argocd/commands/admin/appset.go]: It was meant to avoid polluting the output. Code was removed since I moved app generation to server side

File: cmd/argocd/commands/headless/headless.go (hunk 1 above)
Comment (alexmt) [reply cmd/argocd/commands/headless/headless.go]: Moving embedded repo server to another PR

File: cmd/argocd/commands/headless/headless.go (hunk 2 above)
Comment (alexmt) [reply cmd/argocd/commands/headless/headless.go]: agree, will do in another PR

**Comment by alexmt** [comment]
Updated the PR with new implementation (instead of closing and reopening). The PR introduces `argocd appset generate` command that uses new API to generate applications/get generation error. 

PTAL @agaudreault , @pasha-codefresh

File: cmd/argocd-server/commands/argocd_server.go [hunk 6]
```diff
@@ -142,7 +144,11 @@ func NewCommand() *cobra.Command {
 
 			dynamicClient := dynamic.NewForConfigOrDie(config)
 
-			controllerClient, err := client.New(config, client.Options{})
```
Comment (alexmt) [inline cmd/argocd-server/commands/argocd_server.go:145]: Change is required if user is trying to `generate` appset with non default project (or create with `--dry-run=true` flag)

File: server/applicationset/applicationset.go [hunk 7]
```diff
@@ -366,6 +365,42 @@ func (s *Server) ResourceTree(ctx context.Context, q *applicationset.Application
...
+
+	logs := bytes.NewBuffer(nil)
+	logger := log.New()
+	logger.SetOutput(logs)
+
+	res := &applicationset.ApplicationSetGenerateResponse{}
+
+	apps, err := s.generateApplicationSetApps(ctx, logger.WithField("applicationset", appset.Name), *appset, namespace)
+	if err != nil {
+		res.ErrorLog = fmt.Sprintf("unable to generate Applications of ApplicationSet: %v\n", err)
```
Comment (pasha-codefresh) [inline server/applicationset/applicationset.go:395]: why not return error and build error message on cli , base on this error ? In any case you will need process error on client part, to show reasonable error when request or permissions are incorrect, i think it will be more consistent to do it also when you cant generate applications

**Review by pasha-codefresh — CHANGES_REQUESTED** [review]

**Comment by pasha-codefresh** [comment]
LGTM, just left small comment

File: server/applicationset/applicationset.go (hunk 7 above)
Comment (alexmt) [reply server/applicationset/applicationset.go]: Makes sense. `ErrorLog` just duplicates error - removed it.

**Review by pasha-codefresh — APPROVED** [review]

File: cmd/argocd/commands/applicationset.go [hunk 8]
```diff
@@ -208,6 +210,74 @@ func NewApplicationSetCreateCommand(clientOpts *argocdclient.ClientOptions) *cob
...
+			appsets, err := cmdutil.ConstructApplicationSet(fileUrl)
+			errors.CheckError(err)
+
+			if len(appsets) != 1 {
+				fmt.Printf("Input file must contain one ApplicationSet")
+				os.Exit(1)
+			}
+			appset := appsets[0]
+			if appset.Name == "" {
+				err := fmt.Errorf("Error generating aps for ApplicationSet %s. ApplicationSet does not have Name field set", appset)
```
Comment (agaudreault) [inline cmd/argocd/commands/applicationset.go:241]:
```suggestion
				err := fmt.Errorf("Error generating apps for ApplicationSet %s. ApplicationSet does not have Name field set", appset)
```

**Review by agaudreault — COMMENTED** [review]
Nice! ❤️ Few small comments, nothing major :)

File: server/applicationset/applicationset.go [hunk 9]
```diff
@@ -366,6 +365,40 @@ func (s *Server) ResourceTree(ctx context.Context, q *applicationset.Application
 	return s.buildApplicationSetTree(a)
 }
 
+func (s *Server) Generate(ctx context.Context, q *applicationset.ApplicationSetGenerateRequest) (*applicationset.ApplicationSetGenerateResponse, error) {
```
Comment (agaudreault) [inline server/applicationset/applicationset.go:368]: What about RBAC for this call? It may allow to discover cluster secrets, provate repos, etc. Should we restrict this to people with the "create" rbac?

File: cmd/argocd/commands/applicationset.go [hunk 10]
```diff
@@ -208,6 +210,74 @@ func NewApplicationSetCreateCommand(clientOpts *argocdclient.ClientOptions) *cob
 	return command
 }
 
+// NewApplicationSetGenerateCommand returns a new instance of an `argocd appset generate` command
+func NewApplicationSetGenerateCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
+	var output string
+	command := &cobra.Command{
+		Use:   "generate",
+		Short: "Generate apps of ApplicationSet",
```
[the remark this answers is not in the document]
Comment (alexmt) [reply cmd/argocd/commands/applicationset.go]: thank you, applied!

File: cmd/argocd/commands/applicationset.go (hunk 8 above)
Comment (alexmt) [reply cmd/argocd/commands/applicationset.go]: accepted suggesion

File: cmd/argocd/commands/applicationset.go [hunk 11]
```diff
@@ -208,6 +210,74 @@ func NewApplicationSetCreateCommand(clientOpts *argocdclient.ClientOptions) *cob
...
+
+			switch output {
+			case "yaml", "json":
+				var resources []interface{}
+				for i := range appsList {
+					app := appsList[i]
+					app.APIVersion = arogappsetv1.ApplicationSchemaGroupVersionKind.GroupVersion().String()
+					app.Kind = arogappsetv1.ApplicationSchemaGroupVersionKind.Kind
+					resources = append(resources, app)
+				}
```
[the remark this answers is not in the document]
Comment (alexmt) [reply cmd/argocd/commands/applicationset.go]: This is due to a strange behavior of go k8s client. For some reason it drops apiVersion/kind . We have to do the same workaround in other places. Example: https://github.com/argoproj/argo-cd/blob/7cf5ed06d4f8105812a9430f52419179762350da/cmd/argocd/commands/app_actions.go#L225

File: server/applicationset/applicationset.go (hunk 9 above)
Comment (alexmt) [reply server/applicationset/applicationset.go:368]: Agree - user must have "create" RBAC to see generated apps. I've added a check here: https://github.com/argoproj/argo-cd/blob/4cbb1a9cafecb0efdcfa604e131bd797a16294da/server/applicationset/applicationset.go#L383

**Comment by alexmt** [comment]
@agaudreault done!

**Review by agaudreault — APPROVED** [review]
